contracts/arbitrage/BancorArbitrage.sol

```diff
@@ -0,0 +1,479 @@
+
+            uint256 val = sourceToken.isNative() ? sourceAmount : 0;
+
+            // perform the trade
+            _bancorNetworkV2.convertByPath{ value: val }(
+                path,
+                sourceAmount,
+                minTargetAmount,
+                address(0x0),
+                address(0x0),
+                0
+            );
+
+            return;
+        }
+
+        if (exchangeId == EXCHANGE_ID_BANCOR_V3) {
+            // allow the network to withdraw the source tokens
+            sourceToken.safeApprove(address(_bancorNetworkV3), sourceAmount);
+
+            // perform the trade
+            _bancorNetworkV3.tradeBySourceAmount(
+                sourceToken,
+                targetToken,
+                sourceAmount,
+                minTargetAmount,
+                deadline,
+                address(0x0)
+            );
+
+            return;
+        }
+
+        if (exchangeId == EXCHANGE_ID_UNISWAP_V2 || exchangeId == EXCHANGE_ID_SUSHISWAP) {
+            IUniswapV2Router02 router = exchangeId == EXCHANGE_ID_UNISWAP_V2 ? _uniswapV2Router : _sushiSwapRouter;
+
+            // allow the router to withdraw the source tokens
+            sourceToken.safeApprove(address(router), sourceAmount);
+
+            // build the path
+            address[] memory path = new address[](2);
+
+            // perform the trade
+            if (sourceToken.isNative()) {
+                path[0] = address(router.WETH());
+                path[1] = address(targetToken);
+                router.swapExactETHForTokens{ value: sourceAmount }(minTargetAmount, path, address(this), deadline);
+            } else if (targetToken.isNative()) {
+                path[0] = address(sourceToken);
+                path[1] = address(router.WETH());
+                router.swapExactTokensForETH(sourceAmount, minTargetAmount, path, address(this), deadline);
+            } else {
+                path[0] = address(sourceToken);
+                path[1] = address(targetToken);
+                router.swapExactTokensForTokens(sourceAmount, minTargetAmount, path, address(this), deadline);
+            }
+
+            return;
+        }
+
+        if (exchangeId == EXCHANGE_ID_UNISWAP_V3) {
+            address tokenIn = sourceToken.isNative() ? address(_weth) : address(sourceToken);
+            address tokenOut = targetToken.isNative() ? address(_weth) : address(targetToken);
+
+            if (tokenIn == address(_weth)) {
+                IWETH(address(_weth)).deposit{ value: sourceAmount }();
+            }
+
+            // allow the router to withdraw the source tokens
+            Token(tokenIn).safeApprove(address(_uniswapV3Router), sourceAmount);
+
+            // build the params
+            ISwapRouter.ExactInputSingleParams memory params = ISwapRouter.ExactInputSingleParams({
+                tokenIn: tokenIn,
+                tokenOut: tokenOut,
+                fee: uint24(customInt), // fee
```

Comment:
This is a dangerous (unchecked) cast.
Better use OZ's `SafeCast.toUint24()`.

Reply:
toUint24 isn't supported on SafeCast, but in this case it's not really an issue - an overflow would mean that the function won't find the pool with the specified fee and the tx will revert. and if it does find a pool (practically impossible), it would use a pool with different fee for a sub-optimal arb which will either revert later on (can't return flashloan) or will be a bit less profitable, so no harm done

Reply:
Maybe pass `customInt` as `uint24` then.

Reply:
Can’t - the idea is that it’s a custom field that any exchange can use differently - one might use as uint24 and another might use it as unit256